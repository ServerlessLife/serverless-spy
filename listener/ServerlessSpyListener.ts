import { PrettifyForDisplay } from './PrettifyForDisplay';
import {
  DynamoDBSpyHandler,
  EventBridgeSpyHandler,
  EventBridgeRuleSpyHandler,
  S3SpyHandler,
  SnsSubscriptionSpyHandler,
  SnsTopicSpyHandler,
  SqsSpyHandler,
  FunctionRequestSpyHandler,
  FunctionConsoleSpyHandler,
  FunctionResponseSpyHandler,
  FunctionErrorSpyHandler,
} from './SpyHandlers.ts';
import { WaitForParams } from './WaitForParams';
import { DynamoDBSpyEvent } from '../common/spyEvents/DynamoDBSpyEvent';
import { EventBridgeRuleSpyEvent } from '../common/spyEvents/EventBridgeRuleSpyEvent';
import { EventBridgeSpyEvent } from '../common/spyEvents/EventBridgeSpyEvent';
import { FunctionConsoleSpyEvent } from '../common/spyEvents/FunctionConsoleSpyEvent';
import { FunctionErrorSpyEvent } from '../common/spyEvents/FunctionErrorSpyEvent';
import { FunctionRequestSpyEvent } from '../common/spyEvents/FunctionRequestSpyEvent';
import { FunctionResponseSpyEvent } from '../common/spyEvents/FunctionResponseSpyEvent';
import { S3SpyEvent } from '../common/spyEvents/S3SpyEvent';
import { SnsSubscriptionSpyEvent } from '../common/spyEvents/SnsSubscriptionSpyEvent';
import { SnsTopicSpyEvent } from '../common/spyEvents/SnsTopicSpyEvent';
import { SqsSpyEvent } from '../common/spyEvents/SqsSpyEvent';

export type ServerlessSpyListener<TSpyEvents> = {
  [P in keyof FilterConditionally<TSpyEvents, `DynamoDB#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<DynamoDBSpyEvent>>
    >
  ) => Promise<DynamoDBSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `EventBridge#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<EventBridgeSpyEvent<T>>>
    >
  ) => Promise<EventBridgeSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `EventBridgeRule#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<EventBridgeRuleSpyEvent<T>>>
    >
  ) => Promise<EventBridgeRuleSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `S3#${any}`> &
    string as `waitFor${P}`]: (
    param?: PrettifyForDisplay<WaitForParams<PrettifyForDisplay<S3SpyEvent>>>
  ) => Promise<S3SpyHandler>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `SnsSubscription#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<SnsSubscriptionSpyEvent<T>>>
    >
  ) => Promise<SnsSubscriptionSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `SnsTopic#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<SnsTopicSpyEvent<T>>>
    >
  ) => Promise<SnsTopicSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Sqs#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<SqsSpyEvent<T>>>
    >
  ) => Promise<SqsSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Request`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionRequestSpyEvent<T>>>
    >
  ) => Promise<FunctionRequestSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Console`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionConsoleSpyEvent<T>>>
    >
  ) => Promise<FunctionConsoleSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Error`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionErrorSpyEvent<T>>>
    >
  ) => Promise<FunctionErrorSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Response`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionResponseSpyEvent<T>>>
    > //      filter?: (event: FunctionResponseSpyEvent<T>) => boolean
  ) => Promise<FunctionResponseSpyHandler<T>>;
} & {
  stop: () => void;
};

type FilterConditionally<Source, Condition> = Pick<
  Source,
  {
    [K in keyof Source]: Source[K] extends Condition ? K : never;
  }[keyof Source]
>;
