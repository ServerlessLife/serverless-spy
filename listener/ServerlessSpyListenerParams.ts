import { Credentials } from '@aws-sdk/types';

export type ServerlessSpyListenerParams = {
  credentials?: Credentials;
  debugMode?: boolean;
  scope: string;
};
