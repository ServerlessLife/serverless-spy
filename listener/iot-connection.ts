import { IoTClient, DescribeEndpointCommand } from '@aws-sdk/client-iot';
import { device } from 'aws-iot-device-sdk';

export const SSPY_TOPIC = 'sspy';

export type fragment = { id: string; index: number; count: number; data: any };

function createLog(debugMode: boolean) {
  return (message: string, ...optionalParams: any[]) => {
    if (debugMode) {
      console.debug('SSPY', message, ...optionalParams);
    }
  };
}

function createErrorLog() {
  return (message: string, ...optionalParams: any[]) => {
    console.error('SSPY', message, ...optionalParams);
  };
}

export async function getConnection(debugMode: boolean): Promise<device> {
  const log = createLog(debugMode);
  const logError = createErrorLog();
  log('Getting IoT endpoint');
  let response: any;
  try {
    const iotClient = new IoTClient({});
    response = await iotClient.send(
      new DescribeEndpointCommand({
        endpointType: 'iot:Data-ATS',
      })
    );
  } catch (e) {
    logError('failed to get endpoint', e);
    throw e;
  }
  log('Using IoT endpoint:', response.endpointAddress);

  if (!response.endpointAddress) {
    logError('No IoT endpoint could be found');
    throw new Error('IoT Endpoint address not found');
  }

  const connection = new device({
    protocol: 'wss',
    host: response.endpointAddress,
    region: process.env['AWS_REGION'],
    reconnectPeriod: 1,
  });

  connection.on('connect', () => {
    log('IoT connected');
  });

  connection.on('error', (err) => {
    logError('IoT error', err);
  });

  connection.on('close', () => {
    log('IoT closed');
  });

  connection.on('reconnect', () => {
    log('IoT reconnected');
  });

  return connection;
}
