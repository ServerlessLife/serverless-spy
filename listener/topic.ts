export const SSPY_TOPIC = 'sspy';

export function getTopic(scope: string) {
  return `${SSPY_TOPIC}/${scope}`;
}
