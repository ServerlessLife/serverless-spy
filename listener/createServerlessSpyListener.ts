import { ServerlessSpyListenerParams } from './ServerlessSpyListenerParams';
import { WsListener } from './WsListener';

export async function createServerlessSpyListener<TSpyEvents>(
  params: ServerlessSpyListenerParams
) {
  const wsListener = new WsListener<TSpyEvents>();
  let resolve, reject: ((value: void | PromiseLike<void>) => void) | undefined;
  const promise = new Promise<void>((res, rej) => {
    resolve = res;
    reject = rej;
  });
  await wsListener.start({
    ...params,
    connectionOpenResolve: params.connectionOpenResolve || resolve,
    connectionOpenReject: params.connectionOpenReject || reject,
  });

  const proxy = wsListener.createProxy();
  await promise;
  return proxy;
}
