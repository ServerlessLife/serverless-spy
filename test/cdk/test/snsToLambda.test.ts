import * as fs from 'fs';
import * as path from 'path';
import { PublishCommand, SNSClient } from '@aws-sdk/client-sns';
import { SNSEvent } from 'aws-lambda';
import { v4 as uuidv4 } from 'uuid';
import { TestData } from './TestData';
import { createServerlessSpyListener } from '../../../listener/createServerlessSpyListener';
import { ServerlessSpyListener } from '../../../listener/ServerlessSpyListener';
import { ServerlessSpyEvents } from '../serverlessSpyEvents/ServerlessSpyEventsSnsToLambda';

jest.setTimeout(30000);

describe('SNS to Lambda', () => {
  const exportLocation = path.join(__dirname, '../cdkOutput.json');
  let serverlessSpyListener: ServerlessSpyListener<ServerlessSpyEvents>;

  if (!fs.existsSync(exportLocation)) {
    throw new Error(`File ${exportLocation} does not exists.`);
  }
  const output = JSON.parse(fs.readFileSync(exportLocation).toString())[
    'ServerlessSpySnsToLambda'
  ];

  beforeEach(async () => {
    serverlessSpyListener =
      await createServerlessSpyListener<ServerlessSpyEvents>({
        serverlessSpyWsUrl: output.ServerlessSpyWsUrl,
      });
  });

  afterEach(async () => {
    serverlessSpyListener.stop();
  });

  test('Basic test', async () => {
    const id = uuidv4();
    const data = <TestData>{
      id,
      message: 'Hello',
    };

    const snsClient = new SNSClient({});
    const params = new PublishCommand({
      Message: JSON.stringify(data),
      TopicArn: output.SnsTopicArnMyTopic,
    });
    await snsClient.send(params);

    (
      await serverlessSpyListener.waitForSnsTopicMyTopic<TestData>({
        condition: (d) => d.message.id === id,
      })
    ).toMatchObject({ message: data });

    (
      await serverlessSpyListener.waitForSnsSubscriptionMyTopicMyLambda<TestData>(
        {
          condition: (d) => d.message.id === id,
        }
      )
    ).toMatchObject({ message: data });

    (
      await (
        await serverlessSpyListener.waitForFunctionMyLambdaRequest<SNSEvent>({
          condition: (d) =>
            !!d.request.Records.map(
              (r) => JSON.parse(r.Sns.Message) as TestData
            ).find((d) => d.id === data.id),
        })
      )
        .toMatchObject({
          request: {
            Records: expect.arrayContaining([
              expect.objectContaining({
                Sns: expect.objectContaining({
                  Message: JSON.stringify(data),
                }),
              }),
            ]),
          },
        })
        .followedByResponse({})
    ).toMatchObject({
      request: {
        Records: expect.arrayContaining([
          expect.objectContaining({
            Sns: expect.objectContaining({
              Message: JSON.stringify(data),
            }),
          }),
        ]),
      },
    });

    (
      await (
        await serverlessSpyListener.waitForFunctionMyPythonLambdaRequest<SNSEvent>(
          {
            condition: (d) =>
              !!d.request.Records.map(
                (r) => JSON.parse(r.Sns.Message) as TestData
              ).find((d) => d.id === data.id),
          }
        )
      )
        .toMatchObject({
          request: {
            Records: expect.arrayContaining([
              expect.objectContaining({
                Sns: expect.objectContaining({
                  Message: JSON.stringify(data),
                }),
              }),
            ]),
          },
        })
        .followedByResponse({})
    ).toMatchObject({
      request: {
        Records: expect.arrayContaining([
          expect.objectContaining({
            Sns: expect.objectContaining({
              Message: JSON.stringify(data),
            }),
          }),
        ]),
      },
    });
  });
});
