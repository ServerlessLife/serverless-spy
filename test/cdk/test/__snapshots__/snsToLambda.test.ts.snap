// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNS to Lambda Snapshot 1`] = `
{
  "Outputs": {
    "SnsTopicArnMyTopic": {
      "Value": {
        "Ref": "MyTopic86869434",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyLambdaAllowInvokeTestMyTopic06942AEA77ADE5E7": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "MyLambdaCCE802FB",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MyLambdaCCE802FB": {
      "DependsOn": [
        "MyLambdaServiceRoleDefaultPolicy5BBC6F68",
        "MyLambdaServiceRole4539ECB6",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "798d95ca5ea9167fe4c020efbb7df9d7998800b0911757a1ea86cfe4b8fc2c2f.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_LAMBDA_EXEC_WRAPPER": "/opt/spy-wrapper",
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
            "SSPY_DEBUG": "true",
            "SSPY_FUNCTION_NAME": "MyLambda",
            "SSPY_INFRA_MAPPING": "{}",
            "SSPY_ROOT_STACK": "Test",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "ServerlessSpyExtensionEA4873A0",
          },
          {
            "Ref": "ServerlessSpySpyWrapperLayerBE5627BF",
          },
        ],
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "MyLambdaServiceRole4539ECB6",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyLambdaMyTopicA81E02FD": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "MyLambdaCCE802FB",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "MyLambdaServiceRole4539ECB6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyLambdaServiceRoleDefaultPolicy5BBC6F68": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyLambdaServiceRoleDefaultPolicy5BBC6F68",
        "Roles": [
          {
            "Ref": "MyLambdaServiceRole4539ECB6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyPythonLambdaAllowInvokeTestMyTopic06942AEA71E065BF": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "MyPythonLambdaBD2BE2C6",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MyPythonLambdaBD2BE2C6": {
      "DependsOn": [
        "MyPythonLambdaServiceRoleDefaultPolicyEBB1A269",
        "MyPythonLambdaServiceRoleE254838E",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a34102ca9b461fd6335e146271be5476c61e07b3b0ca21959ff0705dd70432d8.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_LAMBDA_EXEC_WRAPPER": "/opt/spy-wrapper",
            "SSPY_DEBUG": "true",
            "SSPY_FUNCTION_NAME": "MyPythonLambda",
            "SSPY_INFRA_MAPPING": "{}",
            "SSPY_ROOT_STACK": "Test",
          },
        },
        "Handler": "dummy.handler",
        "Layers": [
          {
            "Ref": "ServerlessSpyPythonExtension718A6034",
          },
          {
            "Ref": "ServerlessSpySpyWrapperLayerBE5627BF",
          },
        ],
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "MyPythonLambdaServiceRoleE254838E",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyPythonLambdaMyTopicC8D1E8B1": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "MyPythonLambdaBD2BE2C6",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "MyPythonLambdaServiceRoleDefaultPolicyEBB1A269": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyPythonLambdaServiceRoleDefaultPolicyEBB1A269",
        "Roles": [
          {
            "Ref": "MyPythonLambdaServiceRoleE254838E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyPythonLambdaServiceRoleE254838E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTopic86869434": {
      "Type": "AWS::SNS::Topic",
    },
    "ServerlessSpyExtensionEA4873A0": {
      "Properties": {
        "CompatibleArchitectures": [
          "x86_64",
        ],
        "CompatibleRuntimes": [
          "nodejs16.x",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "babbb172503f6a6ffc178f8d107bae18937392efdbc57c9a729dd35ef6228e53.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "ServerlessSpyPythonExtension718A6034": {
      "Properties": {
        "CompatibleArchitectures": [
          "x86_64",
        ],
        "CompatibleRuntimes": [
          "python3.9",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f0805f3798a1ec284ae50530bcfe1f7dc297a30367d2ef35819e3b21b57db6cd.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "ServerlessSpySpyWrapperLayerBE5627BF": {
      "Properties": {
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "bf2df0858ecef77ac31b1994b46cd25f69a84fea5d58e9b5e5c6e458c33012f6.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "ServerlessSpySubscription0036DF4AB": {
      "DependsOn": [
        "ServerlessSpySubscription0ServiceRoleDefaultPolicyD8FA77F4",
        "ServerlessSpySubscription0ServiceRoleF955D76C",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2fc02cc64fedb8a5c7c3049c58656172132663d36b541073ada7da4078957efa.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
            "SSPY_INFRA_MAPPING": {
              "Fn::Join": [
                "",
                [
                  "{"",
                  {
                    "Ref": "MyTopic86869434",
                  },
                  "":"SnsTopic#MyTopic"}",
                ],
              ],
            },
            "SSPY_ROOT_STACK": "Test",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "ServerlessSpySubscription0ServiceRoleF955D76C",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ServerlessSpySubscription0AllowInvokeTestMyTopic06942AEAB4AEE620": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ServerlessSpySubscription0036DF4AB",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ServerlessSpySubscription0MyTopic191FC9F5": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "ServerlessSpySubscription0036DF4AB",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "ServerlessSpySubscription0ServiceRoleDefaultPolicyD8FA77F4": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServerlessSpySubscription0ServiceRoleDefaultPolicyD8FA77F4",
        "Roles": [
          {
            "Ref": "ServerlessSpySubscription0ServiceRoleF955D76C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ServerlessSpySubscription0ServiceRoleF955D76C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ServerlessSpySubscription13C6B2B7D": {
      "DependsOn": [
        "ServerlessSpySubscription1ServiceRoleDefaultPolicyB2E063F5",
        "ServerlessSpySubscription1ServiceRoleCB639294",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2fc02cc64fedb8a5c7c3049c58656172132663d36b541073ada7da4078957efa.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
            "SSPY_INFRA_MAPPING": {
              "Fn::Join": [
                "",
                [
                  "{"",
                  {
                    "Ref": "MyTopic86869434",
                  },
                  "":"SnsSubscription#MyTopic#MyLambda"}",
                ],
              ],
            },
            "SSPY_ROOT_STACK": "Test",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "ServerlessSpySubscription1ServiceRoleCB639294",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ServerlessSpySubscription1AllowInvokeTestMyTopic06942AEAED481931": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ServerlessSpySubscription13C6B2B7D",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ServerlessSpySubscription1MyTopic98E4131B": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "ServerlessSpySubscription13C6B2B7D",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "ServerlessSpySubscription1ServiceRoleCB639294": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ServerlessSpySubscription1ServiceRoleDefaultPolicyB2E063F5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServerlessSpySubscription1ServiceRoleDefaultPolicyB2E063F5",
        "Roles": [
          {
            "Ref": "ServerlessSpySubscription1ServiceRoleCB639294",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ServerlessSpySubscription2AllowInvokeTestMyTopic06942AEA9C58EED7": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ServerlessSpySubscription2F82BAF5F",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ServerlessSpySubscription2F82BAF5F": {
      "DependsOn": [
        "ServerlessSpySubscription2ServiceRoleDefaultPolicyA3D9E5E8",
        "ServerlessSpySubscription2ServiceRole7162AC7F",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2fc02cc64fedb8a5c7c3049c58656172132663d36b541073ada7da4078957efa.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
            "SSPY_INFRA_MAPPING": {
              "Fn::Join": [
                "",
                [
                  "{"",
                  {
                    "Ref": "MyTopic86869434",
                  },
                  "":"SnsSubscription#MyTopic#MyPythonLambda"}",
                ],
              ],
            },
            "SSPY_ROOT_STACK": "Test",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "ServerlessSpySubscription2ServiceRole7162AC7F",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ServerlessSpySubscription2MyTopic089E6498": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "ServerlessSpySubscription2F82BAF5F",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "ServerlessSpySubscription2ServiceRole7162AC7F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ServerlessSpySubscription2ServiceRoleDefaultPolicyA3D9E5E8": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServerlessSpySubscription2ServiceRoleDefaultPolicyA3D9E5E8",
        "Roles": [
          {
            "Ref": "ServerlessSpySubscription2ServiceRole7162AC7F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
